### PR TITLE
[Model] Mamba2: Add exact replay for bit-identical SSM paths

### PR DESCRIPTION
**Pull request by NolenLiang**
## Purpose

Mamba2 layers run the chunked scan (SSD) for prefill and the recurrent update (SSU) for decode; the two are not bit-identical, so a request's Mamba-layer outputs depend on how it was scheduled. This adds an opt-in, default-off `--mamba-exact-replay` mode in which every SSD call for a sequence starts from the fp32 state at the sequence's last chunk boundary and re-feeds the buffered inputs (x, raw dt, B, C) of the current partial chunk, so prefill, chunked prefill and decode all see the chunk grid of a single-shot prefill and the Mamba2 SSM computation produces identical bits on every path. Bit-identical logits additionally need the batch-invariant kernels for the other layers; in this mode the Mamba2 attention backend is accepted by `VLLM_BATCH_INVARIANT=1`.

Design discussion: #55524. Data behind the design: vllm-project/vllm#54993 (comments), issue #27433.

## Why this is not a duplicate

- #54993 (fixed physical chunking of prefill in the scheduler) and its author's packed-history recovery prototype (#27433, 2026-09-03) make *recovery after preemption reproduce the SSU decode numerics*, which gives scheduling-independent served logprobs at no per-step cost. This PR makes *decode reproduce the SSD prefill numerics*, which additionally gives `prefill(prefix + token) == decode(token)` (the property `tests/v1/determinism/test_batch_invariance.py::test_decode_logprobs_match_prefill_logprobs` checks) — needed for RL training/inference alignment, where the trainer scores generated tokens with the chunked scan. The two are complementary and can coexist; this mode is an independent opt-in.
- The combination with #54993's scheduler change has not been tested; the two are independent options.
- ReplaySSM (`--use-replayssm`) is a decode bandwidth optimisation with its own 16-slot ring; its `dt_cache` holds softplus(dt + bias) and its cursor counts flush steps, so it cannot serve as the replay buffer. The modes are mutually exclusive.

## Changes

- `CacheConfig.mamba_exact_replay` / `--mamba-exact-replay`; `VllmConfig.validate_mamba_exact_replay` fails closed unless: model opts in (`SupportsMambaExactReplay`, currently `Mamba2ForCausalLM`), `mamba_ssm_cache_dtype=float32`, `mamba_cache_mode=none`, Triton mamba backend, no speculative decoding, no ReplaySSM, no KV connectors, TP=1/PP=1, no async scheduling, no DBO / micro-batching, `enforce_eager`.
- Four token-major partial-chunk buffers (`x`, raw `dt`, `B`, `C`; `chunk_size` entries per slot) appended to the mamba state; shapes/dtypes added consistently on the model-config path (`get_mamba_state_{shape,dtype}_from_config`) and the layer path so the page size cannot desync.
- `vllm/model_executor/layers/mamba/exact_replay.py`: `ExactReplayBuffers` and `exact_replay_ssd()`, one function for prefill and decode rows (prepend buffered tokens, run the SSD kernels, write back only this step's outputs, advance the fp32 boundary state only when a chunk completes, store the trailing partial chunk).
- `Mamba2AttentionMetadata.exact_replay_{p,d}` built on the host from `num_computed_tokens` (`build_exact_replay_metadata`); `MambaMixer2` routes both prefill and decode rows through `exact_replay_ssd` when the mode is on.
- Batch-invariance gate: `get_mamba_attn_backend(mamba_type, mamba_exact_replay)` accepts the Mamba2 backend in batch-invariant mode only when the option is set (config-level, no process-global state).
- Docs: `docs/features/batch_invariance.md`, section "Mamba2 Models: Exact-Replay Mode".

## Test plan

Commands run on GB200, on the v0.28.0-based branch and again on this main-based branch (nightly binaries for the base commit):

```
python -m pytest tests/v1/attention/test_mamba2_exact_replay_metadata.py \
    tests/v1/attention/test_mamba2_exact_replay_config.py \
    tests/kernels/mamba/test_mamba2_exact_replay.py
python -m pytest tests/models/test_registry.py -k exact_replay
```

- Metadata unit tests: two exact layouts (decode rows mid-chunk / at boundary / completing a chunk; fresh and unaligned-resumed prefill rows) plus 20 random layouts checked against structural invariants.
- Config tests: the supported configuration is accepted; fp32 SSM cache, no prefix caching, no ReplaySSM, no async scheduling and no DBO are enforced; the selector accepts the Mamba2 backend under `VLLM_BATCH_INVARIANT=1` only with the option; the registry reports the opt-in for `Mamba2ForCausalLM` only.
- Kernel-level test: four sequences of different lengths, first prefill split at unaligned points, resumed, then interleaved decode until each finishes (batch composition and chunk offsets differ every step); every output row equals the single-shot prefill row bitwise; run with contiguous buffers and with buffers carved from paged storage the way the mamba cache is.
- Engine-level (script outside the repo, `AntonV/mamba2-130m-hf`, 24 Mamba2 layers, bf16, fp32 SSM cache, eager, TP1, one request, random 900-token stream; full raw-logit rows compared with `torch.equal` against a single-shot prefill):
  - baseline: forced token-by-token decode differs in 299/299 rows; chunked prefill (`max_num_batched_tokens=100`) differs in 800/900 rows;
  - exact replay: 0/299 and 0/900 rows differ, with and without `VLLM_BATCH_INVARIANT=1`; single-shot prefill logits unchanged by the mode; `VLLM_BATCH_INVARIANT=1` without the option fails closed at startup.
- Cost (launch-bound, eager, batch 1): 300 decode steps 6.5 s → 11.9 s (≈ +0.75 ms per layer per step); memory for this model 0.887 MiB buffers + 0.750 MiB fp32 state per layer per request (≈ 39.3 MiB per request over 24 layers).

Not yet covered (WIP): real scheduler preemption on a hybrid model, multi-request mixed-offset decode at the engine level (kernel-level test only), TP > 1, CUDA graphs, spec decode, prefix caching, performance beyond the launch-bound setting.

## AI assistance

AI tools assisted with the implementation, tests and documentation. I reviewed every changed line and ran the tests above.

### CODE REVIEW

**Comment by aoshen02** [comment]
You might need to simplify the code and split it into kernel-wise batch invariance pr.

**Comment by NolenLiang** [comment]
Closing this PR. Following the review, the change is being redone at kernel granularity: no separate option and no per-model opt-in; batch-invariant mode drives it for every Mamba2 model. The replacement PR follows and will be linked here.

**Comment by NolenLiang** [comment]
Replaced by #55627.
